SmartDeviceLink/SDLSecondaryTransportManager.m

```diff
@@ -0,0 +1,715 @@
+    __weak typeof(self) weakSelf = self;
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(RetryConnectionDelay * NSEC_PER_SEC)), _stateMachineQueue, ^{
+        if ([weakSelf.stateMachine isCurrentState:SDLSecondaryTransportStateReconnecting]) {
+            SDLLogD(@"Retry secondary transport after disconnection or registration failure");
+            [weakSelf.stateMachine transitionToState:SDLSecondaryTransportStateConfigured];
+        }
+    });
+}
+
+- (void)sdl_startManager {
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_onAppStateUpdated:) name:UIApplicationDidBecomeActiveNotification object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_onAppStateUpdated:) name:UIApplicationWillResignActiveNotification object:nil];
+
+    [self.primaryProtocolHandler start];
+}
+
+- (void)sdl_stopManager {
+    SDLLogD(@"SDLSecondaryTransportManager stop");
+
+    [[NSNotificationCenter defaultCenter] removeObserver:self name:UIApplicationDidBecomeActiveNotification object:nil];
+    [[NSNotificationCenter defaultCenter] removeObserver:self name:UIApplicationWillResignActiveNotification object:nil];
+
+    [self.primaryProtocolHandler stop];
+
+    self.streamingServiceTransportMap = [@{@(SDLServiceTypeAudio):@(SDLTransportClassInvalid),
+                                @(SDLServiceTypeVideo):@(SDLTransportClassInvalid)} mutableCopy];
+    self.secondaryTransportType = SDLSecondaryTransportTypeDisabled;
+    self.transportsForAudioService = @[];
+    self.transportsForVideoService = @[];
+
+    self.ipAddress = nil;
+    self.tcpPort = TCPPortUnspecified;
+}
+
+- (void)sdl_configureManager:(nullable NSArray<SDLSecondaryTransportTypeBox *> *)availableSecondaryTransports
+          availableTransportsForAudio:(nullable NSArray<SDLTransportClassBox *> *)availableTransportsForAudio
+          availableTransportsForVideo:(nullable NSArray<SDLTransportClassBox *> *)availableTransportsForVideo {
+    if (![self.stateMachine isCurrentState:SDLSecondaryTransportStateStarted]) {
+        SDLLogW(@"SecondaryTransportManager ignores duplicate Start Service ACK frame");
+        return;
+    }
+
+    SDLSecondaryTransportType secondaryTransportType = SDLSecondaryTransportTypeDisabled;
+    if (availableSecondaryTransports == nil || availableSecondaryTransports.count == 0) {
+        SDLLogW(@"Did not receive secondary transport type from system. Secondary transport is disabled.");
+        secondaryTransportType = SDLSecondaryTransportTypeDisabled;
+    } else {
+        // current proposal says the list should contain only one element
+        SDLSecondaryTransportTypeBox *transportType = availableSecondaryTransports[0];
+        secondaryTransportType = [transportType integerValue];
+    }
+
+    SDLSecondaryTransportType primaryTransportType = [self sdl_getTransportTypeFromProtocol:self.primaryProtocol];
+    if (primaryTransportType == secondaryTransportType) {
```

Comment:
I think this `if` can be `if (primaryTransportType == secondaryTransportType || primaryTransportType == SDLSecondaryTransportTypeDisabled)` and the below `if`s removed.

Reply:
Hmm, unfortunately we cannot combine the logics. In the case of `primaryTransportType == secondaryTransportType` we set `secondaryTransportType` to `SDLSecondaryTransportTypeDisabled`, while in the case of `availableTransportsForAudio == nil` we won't update `secondaryTransportType`. Besides, `primaryTransportType` will not be `SDLSecondaryTransportTypeDisabled`.

I agree that intializing `availableTransportsForAudio` multiple times looks redundant, so I tweaked the logic in commit 068f8c142989fb8035fffed02f9c5d818b25d8cb. The default values are set in the first place, then they will be overwritten. Hope it will make the code cleaner.